src/audio/copier/copier_gain.h

```diff
@@ -0,0 +1,136 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2024 Intel Corporation. All rights reserved.
+ *
+ * Author: Ievgen Ganakov <ievgen.ganakov@intel.com>
+ */
+
+#ifndef __SOF_COPIER_GAIN_H__
+#define __SOF_COPIER_GAIN_H__
+
+#include <sof/audio/buffer.h>
+#include <xtensa/tie/xt_hifi3.h>
+#include <ipc4/base_fw.h>
+#include <ipc/dai.h>
+
+/**
+ * @file copier_gain.h
+ * @brief Header file containing definitions and functions related to audio gain
+ * processing for a copier module.
+ *
+ * This file provides functions, constants and structure definitions for applying gain to
+ * input audio buffers, both in 16-bit and 32-bit container formats. The gain can be
+ * applied in different directions (addition or subtraction) and has three modes:
+ * - static gain
+ * - transition gain (fade-in/fade-out)
+ * - transition mute.
+ */
+
+#define MAX_GAIN_COEFFS_CNT 4 /**< Maximum number of gain coefficients */
+
+/* Common const values for applying gain feature */
+#define Q10_TO_Q31_SHIFT 6
+#define Q10_TO_Q15_SHIFT 5
+
+/* 16x2 store operation requires shift to middle part of 32 bit register */
+#define I64_TO_I16_SHIFT  48
+#define MIDDLE_PART_SHIFT 8
+
+/* Unit gain in q10 format applied by default */
+#define UNITY_GAIN_4X_Q10  0x0400040004000400ULL
+#define UNITY_GAIN_GENERIC 0x400
+#define MAX_INT64          0x7FFFFFFFFFFFFFFFULL
+
+/* Default volume transition delays */
+#define GAIN_DEFAULT_HQ_TRANS_MS     500
+#define GAIN_DEFAULT_SPEECH_TRANS_MS 100
+#define GAIN_ZERO_TRANS_MS           0xFFFF
+
+struct dai_data;
+
+/**
+ * @brief Enumeration representing the state of the copier gain processing.
+ */
+enum copier_gain_state {
+	TRANS_MUTE = 0, /**< Transition to mute state */
+	TRANS_GAIN = 1, /**< Transition to gain state */
+	STATIC_GAIN = 2, /**< Static gain state */
+};
+
+/**
+ * @brief Enumeration representing the direction of the copier gain processing.
+ */
+enum copier_gain_direction {
+	ADDITION = 0,	/**< Addition gain direction */
+	SUBTRACTION = 1 /**< Subtraction gain direction */
```

Comment:
what does this mean? amplification or attenuation make more sense to me, I don't know what addition and substraction mean.

And usually a gain is signed, so why do we need an additional concept?